### PR TITLE
Fixes Issue #369 by updating the FLI interface

### PR DESCRIPTION
**Pull request by leftink**
Here are updates to the FLI interface to support the latest GPI updates.  I know there are still some issues with iteration through this interface, but it gets the interface mostly there. The changes were significant enough that I wanted to get what I had incorporated into the master branch before fixing issues to get all tests passing.

@stuarthodgson I moved away from using the "force" signal function to make things easier for the different data types.  In my testing, this approach behave the same way as the other way so let me know if there is an issue with my approach.

@chiggs and @stuarthodgson I have some more testing (corner cases) that I am pretty sure will have some issues that I will need some support from you guys for how to handle properly.  As an example, when iterating, assume there is a std_logic_vector signal name "test_sig" under the dut.  When iterating over "/dut" there will be an attribute "/dut/test_sig".  If we then iterate over "test_sig", there would be signals with the following names "/dut/test_sig(0)", "/dut/test_sig(1)", etc.  If you walk through how handles.py will handle this in discover_all, it looks to me the handles will be cached as follows: "dut.test_sig.test_sig[0]", "dut.test_sigf.test_sig[1]", etc. This doesn't seem right to me, but it may be the way I am handling iteration.

### CODE REVIEW

**Comment by leftink** [comment]
@chiggs and @stuarthodgson 
I am not sure what is going on with the code-quality/landscape.  I clicked on details and don't even see this branch there.  I had this happen once before, but I had to push an additional change and with the update it kicked off and passed.  I don't see away to re-submit this to try and kick off the code-quality check and I want to get this incorporated as soon as possible so I can continue working other issues that are dependent on these changes.

**Comment by stuarthodgson** [comment]
@leftink I should get around to looking over this tonight once home from the day job, don't worry about the landscape bot. It's pretty flaky

File: lib/fli/FliCbHdl.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
+/******************************************************************************
+* Copyright (c) 2014 Potential Ventures Ltd
```
Comment (stuarthodgson) [inline lib/fli/FliCbHdl.cpp]: Change to 2015/16

File: lib/fli/FliImpl.cpp [hunk 2]
```diff
@@ -107,11 +107,179 @@ void FliImpl::sim_end(void)
...
+
+        return true;
+    }
+
+    return false;
+}
+
+bool FliImpl::isValueChar(mtiTypeIdT type)
+{
+    return (mti_TickLength(type) == 256);
```
Comment (stuarthodgson) [inline lib/fli/FliImpl.cpp]: Does 256 have a define or a more meaningful value than 256?

File: lib/fli/FliImpl.cpp [hunk 3]
```diff
@@ -107,11 +107,179 @@ void FliImpl::sim_end(void)
...
+}
+
+bool FliImpl::isValueBoolean(mtiTypeIdT type)
+{
+    if (mti_TickLength(type) == 2) {
+        char **enum_values = mti_GetEnumValues(type);
+        std::string strFalse = enum_values[0];
```
Comment (stuarthodgson) [inline lib/fli/FliImpl.cpp]: Is it guaranteed that enum_values[0] will be "false" and not "true"?

File: lib/fli/FliImpl.cpp [hunk 4]
```diff
@@ -107,11 +107,179 @@ void FliImpl::sim_end(void)
...
+    }
+
+    if (NULL == new_obj) {
+        LOG_DEBUG("Didn't find anything named %s", fq_name.c_str());
+        return NULL;
+    }
+
+    if (new_obj->initialise(name,fq_name) < 0) {
+        LOG_ERROR("Failed to initialise the handle %s", name.c_str());
+        return NULL;
```
Comment (stuarthodgson) [inline lib/fli/FliImpl.cpp]: Delete new_obj?

File: lib/fli/FliImpl.cpp [hunk 5]
```diff
@@ -159,7 +330,36 @@ GpiObjHdl*  FliImpl::native_check_create(std::string &name, GpiObjHdl *parent)
  */
 GpiObjHdl*  FliImpl::native_check_create(uint32_t index, GpiObjHdl *parent)
 {
-    return NULL;
+    if (parent->get_type() == GPI_MODULE or parent->get_type() == GPI_ARRAY) {
+        char buff[15];
```
Comment (stuarthodgson) [inline lib/fli/FliImpl.cpp]: Is their a significance to 15?

File: lib/fli/FliImpl.h [hunk 6]
```diff
@@ -227,14 +346,45 @@ class FliTimerCache {
...
+
+private:
+    void populate_handle_list(OneToMany childType);
+
+private:
+    static GpiIteratorMapping<int, OneToMany> iterate_over;      /* Possible mappings */
+    std::vector<OneToMany> *selected;                            /* Mapping currently in use */
+    std::vector<OneToMany>::iterator one2many;
+
+    std::vector<void *> m_vars;
```
Comment (stuarthodgson) [inline lib/fli/FliImpl.h]: Can a concrete type be used here rather than void?
And do the handles supplied via populate_handle_list need to be released/freed when the iterator is destroyed?

File: lib/fli/FliObjHdl.cpp [hunk 7]
```diff
@@ -0,0 +1,530 @@
...
+                m_ascending   = (mti_TickDir(m_val_type) == 1);
+                m_value_enum  = mti_GetEnumValues(elemType);
+                m_num_enum    = mti_TickLength(elemType);
+                m_num_elems   = mti_TickLength(m_val_type);
+                buff_len      = (m_num_elems + (sizeof(*m_mti_buff) - 1)) / sizeof(*m_mti_buff);
+
+                m_mti_buff    = (mtiInt32T*)malloc(sizeof(*m_mti_buff) * buff_len);
+                if (!m_mti_buff) {
+                    LOG_CRITICAL("Unable to alloc mem for value object mti read buffer: ABORTING");
+                    return -1;
```
Comment (stuarthodgson) [inline lib/fli/FliObjHdl.cpp]: LOG_CRITICAL will not return, see gpi_logging.h

File: lib/fli/FliObjHdl.cpp [hunk 8]
```diff
@@ -0,0 +1,530 @@
...
+    } else {
+        value = mti_GetSignalValue(get_handle<mtiSignalIdT>());
+    }
+
+    return (long)value;
+}
+
+int FliIntObjHdl::set_signal_value(const long value)
+{
+    if (m_const) {
```
Comment (stuarthodgson) [inline lib/fli/FliObjHdl.cpp]: Python does not allow a _setcachedvalue on a ConstantObject so these checks may be redundant.

File: lib/fli/FliObjHdl.cpp [hunk 9]
```diff
@@ -0,0 +1,530 @@
...
+
+    return 0;
+}
+
+int FliRealObjHdl::initialise(std::string &name, std::string &fq_name)
+{
+    int buff_len = 0;
+
+    m_num_elems   = 1;
+    buff_len      = (sizeof(double) + (sizeof(*m_mti_buff) - 1)) / sizeof(*m_mti_buff);
```
Comment (stuarthodgson) [inline lib/fli/FliObjHdl.cpp]: Confused by this.

File: lib/fli/FliObjHdl.cpp [hunk 10]
```diff
@@ -0,0 +1,530 @@
...
+        LOG_CRITICAL("Unable to alloc mem for value object mti read buffer: ABORTING");
+        return -1;
+    }
+ 
+    return FliValueObjHdl::initialise(name, fq_name);
+}
+
+double FliRealObjHdl::get_signal_value_real(void)
+{
+    double *iter = (double *)m_mti_buff;
```
Comment (stuarthodgson) [inline lib/fli/FliObjHdl.cpp]: Would it make more sense the redefine m_mti_buff as the correct type in these sub classes of FliValueObjHandle, it would mean slightly more typing but be clearer and remove casting between types.

File: tests/designs/uart2bus/Makefile [hunk 11]
```diff
@@ -27,10 +27,10 @@ ifeq ($(SIM),aldec)
     GPI_EXTRA=vhpi
 endif
 ifeq ($(SIM),modelsim)
-    GPI_EXTRA=fli
```
Comment (stuarthodgson) [inline tests/designs/uart2bus/Makefile]: I'd expect the Makefile.questa to change to detech GPI_IMPL=fli (which it does not do at the moment erroneously) and keep the design Makefiles as they are.

File: lib/fli/FliCbHdl.cpp (hunk 1 above)
Comment (leftink) [inline lib/fli/FliCbHdl.cpp]: @stuarthodgson Agreed, and also updating FliObjHdl.cpp since it is new as well.

File: lib/fli/FliImpl.cpp (hunk 2 above)
Comment (leftink) [inline lib/fli/FliImpl.cpp]: @stuarthodgson There is no define, this is just how many enum values are defined for a type CHARACTER in the VHDL LRM.  I will add a static const int to the function to give the a more descriptive meaning.

File: lib/fli/FliImpl.cpp (hunk 3 above)
Comment (leftink) [inline lib/fli/FliImpl.cpp]: @stuarthodgson To comply with the VHDL LRM, yes.  However, I do need to update the tests below because the compare values need to be uppercase.

File: lib/fli/FliImpl.cpp (hunk 4 above)
Comment (leftink) [inline lib/fli/FliImpl.cpp]: @stuarthodgson You are correct.

File: lib/fli/FliImpl.cpp (hunk 5 above)
Comment (leftink) [inline lib/fli/FliImpl.cpp]: @stuarthodgson Not really, just needed to be big enough to handle the largest unsigned int, i.e. 4294967295.  That will require 10 characters, plus 3 additional characters for "(", ")" and "\0".  I just preferred it to be a little larger.  If you want I can change it.

File: lib/fli/FliImpl.h (hunk 6 above)
Comment (leftink) [inline lib/fli/FliImpl.h]: @stuarthodgson 
I originally used the actual types, but the issue is each one (m_vars, m_sigs, and m_regs) will each be a unique type since the FLI interface likes to have a different handle type for the different kinds and that would then require a different "m_iterator" for each specific type.  This was the cleanest way to not have to test in the next_handle() function which iterator to use.

File: lib/fli/FliObjHdl.cpp (hunk 7 above)
Comment (leftink) [inline lib/fli/FliObjHdl.cpp]: @stuarthodgson This is just a style I tend to follow, so when looking at the code, even if you don't know what LOG_CRITICAL will do you can tell it will exit at that point.

I can remove it if you want, just let me know.

File: lib/fli/FliObjHdl.cpp (hunk 8 above)
Comment (leftink) [inline lib/fli/FliObjHdl.cpp]: @stuarthodgson That is fine, I will double check and if they can't be set in handle.py, then I will remove these checks

File: lib/fli/FliObjHdl.cpp (hunk 9 above)
Comment (leftink) [inline lib/fli/FliObjHdl.cpp]: @stuarthodgson This is just a trick to ensure the buffer is always a multiple of mtiInt32T (rounded up).

This example is boring:
(sizeof(double) + (sizeof(*m_mti_buff)-1))/sizeof(*m_mti_buff)
(8                     + (4                            -1))/4                                       = 2

Consider a string of characters of length 6 characters.  To store in the m_mti_buff you will need at least 6 bytes, but it needs to be in multiples of m_mti_buff so:

(len*sizeof(char) + (sizeof(*m_mti_buff)-1))/sizeof(*m_mti_buff)
(6   * 1                + (4                            -1) / 4            =        (6+3)/4  = 2 due to all types being integers

File: lib/fli/FliObjHdl.cpp (hunk 10 above)
Comment (leftink) [inline lib/fli/FliObjHdl.cpp]: @stuarthodgson I could make that update.  Unfortunately, it won't get rid of casting completely because for the "enum" type the buffer type is dependent on the number of enums.

File: tests/designs/uart2bus/Makefile (hunk 11 above)
Comment (leftink) [inline tests/designs/uart2bus/Makefile]: @stuarthodgson I submitted issue #372, to address this.  I would prefer not to change the questa makefile at this time because we need to determine all makefiles that need to be updated to support this.  Further in that issue I have proposed a solution that will remove the need for this all together.

**Comment by leftink** [comment]
@stuarthodgson Comments have been addressed either above or with the latest commits.  Let me know if you have any more issues.

File: lib/fli/FliImpl.h (hunk 6 above)
Comment (stuarthodgson) [inline lib/fli/FliImpl.h]: Think this got missed

And do the handles supplied via populate_handle_list need to be released/freed when the iterator is destroyed?

Comment (leftink) [inline lib/fli/FliImpl.h]: @stuarthodgson 
Sorry about that, no the handles don't need to be freed.  Getting the handles through the iterators is just like getting the handles through the "mti_find*()" function.  I think this a pointer's to the actual "handle" that the simulator uses and not a copy of the information.  The approach here is identical to the examples in the FLI documentation

**Comment by stuarthodgson** [comment]
@leftink I'm happy to merge this from a technical standpoint. 

Can I just ask, do all the tests that did pass with FLI still pass, we are unable to test FLI ourselves currently.

**Comment by leftink** [comment]
@stuarthodgson 
Before I started making any changes, I tried running some tests and there was an issue where none of the FLI tests would actually work.  It had something to do with the root object and the default GPI_TYPE being used.  Basically, all FLI tests had an exception and wouldn't run.  The first commit in this series addressed that problem.  So more tests are working now than before this tests, but I can't speak to how many tests are passing compared to to before all the changes to the gpi infrastructure.

I did run multiple tests and got them to pass.  I know there are some tests that are still failing, related to iterating that I plan on addressing next.

My plan after this was incorporated was to focus on getting all tests to pass with the FLI.  Since nothing was working, I didn't want to wait to address everything before getting this code incorporated and with it being a pretty big overhaul of the existing code I didn't want to keep working it without getting some feedback.
